### PR TITLE
[add-fire] Add LSTM to SAC, LSTM fixes and initializations

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Adds LSTM support to the SAC optimizer. Runs, but still testing to see if it trains. 

Caveats: The Q network's memories are still 0'ed out (same as in TF) since we don't run them during inference. 

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by ervteng** [comment]
The simple_rl tests pass, but Hallway doesn't train reliably. Seems like we'll need to add previous action to get it to train.

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 1]
```diff
@@ -85,18 +54,17 @@ def get_trajectory_value_estimates(
...
-        value_estimates = self.policy.actor_critic.critic_pass(
-            vector_obs, visual_obs, memory
+        value_estimates, next_memory = self.policy.actor_critic.critic_pass(
+            vector_obs, visual_obs, memory, sequence_length=batch.num_experiences
         )
 
-        next_value_estimate = self.policy.actor_critic.critic_pass(
-            next_obs, next_obs, next_memory
+        next_value_estimate, _ = self.policy.actor_critic.critic_pass(
+            next_obs, next_obs, next_memory, sequence_length=1
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:67]: Does this work? it seems that critic_pass takes a vector obs and a visual obs but here we feed the same obs as both visual and vector.

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 2]
```diff
@@ -284,3 +284,13 @@ def get_probs_and_entropy(
         else:
             all_probs = torch.cat(all_probs_list, dim=-1)
         return log_probs, entropies, all_probs
+
+    @staticmethod
+    def masked_mean(tensor: torch.Tensor, masks: torch.Tensor) -> torch.Tensor:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:289]: A tiny unit test for it? Making sure that if all the masks are 0 it does not 💥

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 3]
```diff
@@ -369,12 +393,30 @@ def update(self, batch: AgentBuffer, num_sequences: int) -> Dict[str, float]:
...
+
+        if len(memories_list) > 0:
+            memories = torch.stack(memories_list).unsqueeze(0)
+            next_memories = torch.stack(next_memories_list).unsqueeze(0)
+        else:
+            memories = None
+            next_memories = None
+        # Q network memories are 0'ed out, since we don't have them during inference.
+        q_memories = torch.zeros(
+            (memories.shape[0], memories.shape[1], memories.shape[2] // 2)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:417]: Is there no way to use zeros_like here?

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:67]: Welp, that's a bug 🙂

Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py]: Fixed

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:289]: It will totally 💥, good call

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:289]: Added test and fixed 💥 issue

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:417]: Changed to use `zeros_like(next_memories)` which are the same size.
